---
title: "R Notebook"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 
# Tue Jun  7 19:35:18 2022 ------------------------------


```{r}
source("tianfengRwrappers.R")
```

# rat Frontier *GSE174098* carotid
```{r fig.width=4, fig.height=8}
rat10x <- CreateSeuratObject(Read10X("./rat_scRNAseq/"), names.field = 2, names.delim = "-",
                                   project = "rat", min.cells = 10, min.features = 300) %>%
  PercentageFeatureSet(pattern = "^Mt-", col.name = "percent.mt") 

table(rat10x$orig.ident)
VlnPlot(rat10x,"nCount_RNA") /
VlnPlot(rat10x,"percent.mt") /
VlnPlot(rat10x, "nFeature_RNA")


rat10x <- rat10x %>% subset(subset = nFeature_RNA > 400 & nFeature_RNA < 4000 &
                              nCount_RNA > 1000 &  nCount_RNA < 30000 & percent.mt< 10) %>%
    SCTransform(vars.to.regress = "percent.mt", verbose = F) %>% 
    RunPCA() %>% FindNeighbors(dims = 1:20) %>% 
    RunUMAP(dims = 1:20) %>% 
    FindClusters(resolution = 0.1)

```


```{r fig.width=4, fig.height=3}
umapplot(rat10x)
multi_featureplot(c("Acta2","Cnn1","Fn1","Lum","Vcam1","Ly6a","Dlx5","Dlx6","Sost"), rat10x)


f("Bmpr1b", rat10x) /
f("Bmpr1a", rat10x)
table(rat10x$seurat_clusters)

rat10x0 <- subset(rat10x,ident = 0)

f("Sost",rat10x)
ncol(subset(rat10x0, subset = Bmpr1a > 0))
ncol(subset(rat10x0, subset = Bmpr1b > 0))
ncol(subset(rat10x0, subset = Bmpr1a > 0 & Bmpr1b > 0))

ncol(subset(rat10x0, subset = Sost > 0 & Bmpr1b > 0))
ncol(subset(rat10x0, subset = Sost > 0 & Bmpr1a > 0))

ncol(rat10x0)
```


```{r}
mouse10x <- CreateSeuratObject(Read10X("./celldiscovery_mouse_10x/1/"), names.field = 2,
                               names.delim = "-",project = "mouse1", min.cells = 10,
                               min.features = 300) %>%
  PercentageFeatureSet(pattern = "^mt-", col.name = "percent.mt") 

mouse10x2 <- CreateSeuratObject(Read10X("./celldiscovery_mouse_10x/2/"), names.field = 2, names.delim = "-",
                                   project = "mouse2", min.cells = 10, min.features = 300) %>%
  PercentageFeatureSet(pattern = "^mt-", col.name = "percent.mt") 

mouse10x <- merge(mouse10x,mouse10x2)

VlnPlot(mouse10x,"nCount_RNA") /
VlnPlot(mouse10x,"percent.mt") /
VlnPlot(mouse10x, "nFeature_RNA")

mouse10x <- mouse10x %>% subset(subset = nFeature_RNA > 500 & nFeature_RNA < 5000 &
                              nCount_RNA > 1000 &  nCount_RNA < 30000 & percent.mt < 10) %>%
    SCTransform(vars.to.regress = "percent.mt", verbose = F) %>% 
    RunPCA() %>% FindNeighbors(dims = 1:20) %>% 
    RunUMAP(dims = 1:20) %>% 
    FindClusters(resolution = 0.1)
# table(mouse10x$orig.ident)


umapplot(mouse10x)
f("Lmo2",mouse10x)
```


```{r}
SMC2_marker <- as.character(read.table("SMC2")$V1)
library(homologene)

mmSMC2_marker <- homologene(SMC2_marker, inTax = 9606, outTax = 10090)
mmSMC2_marker <- mmSMC2_marker$`10090`
mmSMC2_marker <- intersect(rownames(mouse10x), mmSMC2_marker)
mouse10x <- AddModuleScore(mouse10x, list(mmSMC2_marker), name = "SMC2_score")
f("SMC2_score1",mouse10x, min.cutoff = 0)
Dotplot("SMC2_score1",mouse10x)
```


# mouse coronary *GSE131778*
```{r}
mouse_coronary_countmatrix <- read.csv("./GSE131776_mouse_scRNAseq.txt", sep = "\t")
func <- function(s) {
  paste0(strsplit(s, ".", fixed = T)[[1]][2], "_", strsplit(s, ".", fixed = T)[[1]][1])
}
colnames(mouse_coronary_countmatrix) <- lapply(colnames(mouse_coronary_countmatrix), func) # 拆分样本
```

```{r fig.width= 4, fig.height=8}
mousecor <- CreateSeuratObject(counts = mouse_coronary_countmatrix,
                                   project = "mouse_cor", min.cells = 10, min.features = 300) %>%  PercentageFeatureSet(pattern = "^mt-", col.name = "percent.mt") 

# saveRDS(mousecor,"mousecor.rds")
table(mousecor$orig.ident)
VlnPlot(mousecor,"nCount_RNA") /
VlnPlot(mousecor,"percent.mt") /
VlnPlot(mousecor, "nFeature_RNA")


mousecor <- mousecor %>% subset(subset = nFeature_RNA > 400 & nFeature_RNA < 4000 &
                              nCount_RNA > 1000 &  nCount_RNA < 30000 & percent.mt < 10) %>%
    SCTransform(vars.to.regress = "percent.mt", verbose = F) %>% 
    RunPCA() %>% FindNeighbors(dims = 1:20) %>% 
    RunUMAP(dims = 1:20) %>% 
    FindClusters(resolution = 0.1)
table(mousecor$orig.ident)
saveRDS(mousecor,"mousecor.rds")
```

### SMC2
```{r fig.width=4, fig.height=3}
mousecor <- readRDS("mousecor.rds")
f("Prdm16",mousecor, label.size = 7) + theme(legend.text = element_text(size = 20))
```


```{r fig.width=4, fig.height=3}
mousecor_stromal <- subset(mousecor,idents = c(0,1,2))

mousecor_stromal <- mousecor_stromal %>% RunPCA() %>% FindNeighbors(dims = 1:20) %>% 
    RunUMAP(dims = 1:20) %>% FindClusters(resolution = 0.2)
mousecor_stromal <- mousecor_stromal %>% FindClusters(resolution = 0.2)


mousecor_stromal <- readRDS("mousecor_stromal.rds")

umapplot(mousecor_stromal, label.size = 6) + theme(legend.text = element_text(size = 20))
multi_featureplot(c("Acta2","Cnn1","Fn1","Lum","Vcam1","Ly6a","Dlx5","Dlx6"), mousecor_stromal,labels = NA)

levels(Idents(mousecor_stromal)) <-
  c("SMC1","Fibroblast1","Fibromyocyte","SMC1","Fibroblast2","SMC2")
f("Acta2",mousecor_stromal,label.size = 6)
p <- multi_featureplot(c("Acta2","Cnn1","Fn1","Lum","Vcam1","Ly6a","Dlx5","Dlx6","Lgals3"),mousecor_stromal,labels = NA,label.size = 6)
ggsave("natmed_mouse_coronary_stromal2.png",plot = p, height = 12, width = 16,device = png)
umapplot(mousecor_stromal, label.size = 6) + theme(legend.text = element_text(size = 18))

mousecor_stromal <- AddModuleScore(mousecor_stromal, list(mmSMC2_marker), name = "SMC2_score")
f("SMC2_score1", mousecor_stromal,min.cutoff = 0)

```

# mouse carotid scRNAseq *GSE155513*
### dataload
```{r}
count_mats <- list.files("./GSE155513_RAW/")
count_mats <- count_mats[count_mats != "sampleinfo.txt"]
allList <- lapply(count_mats, function(file) {
  dd <- read.table(paste0("./GSE155513_RAW/", file), row.names = 1,stringsAsFactors = F)
  colnames(dd) <- as.character(dd['gene',])
  dd <- dd[-1,]
  CreateSeuratObject(
    counts = dd,
    project = file, min.cells = 10, min.features = 300
  )
})
# 合并seurat对象
mouse_carotid <- merge(allList[[1]], 
  y = allList[-1], add.cell.ids = count_mats,
  project = "mouse_carotid"
)
rm(allList)

# saveRDS(mouse_carotid,"mouse_carotid.rds")
```

### process
```{r}
mouse_carotid <- readRDS("mouse_carotid.rds")

mouse_carotid <- mouse_carotid %>%  PercentageFeatureSet(pattern = "^mt-", col.name = "percent.mt") %>%
subset(subset = nFeature_RNA > 500 & nFeature_RNA < 3000 &
                              nCount_RNA > 1000 &  nCount_RNA < 20000 & percent.mt < 10) %>%
    SCTransform(vars.to.regress = "percent.mt", verbose = F) %>% 
    RunPCA() %>% FindNeighbors(dims = 1:20) %>% 
    RunUMAP(dims = 1:20) %>% 
    FindClusters(resolution = 0.1)

saveRDS(mouse_carotid,"mouse_carotid.rds")
```

### SMC
```{r}
mouse_carotid <- readRDS("mouse_carotid.rds")
umapplot(mouse_carotid)
umapplot(mouse_carotid,group.by = "orig.ident",label = F)
f("Prdm16",mouse_carotid) #SMC
f("Ly6a",mouse_carotid) #SEM-like cells

multi_featureplot(c("Bmp2","Bmp4","Bmp6"),mouse_carotid)

## BMP4 在这里EC的*大部分*中表达，而在人类样本中BMP4+ EC细胞是少数的
markers <- FindAllMarkers(mouse_carotid,logfc.threshold = 0.5,min.diff.pct = 0.2, only.pos = T)
```

### stromal cells
```{r fig.width=4, fig.height=3}
mouse_carotid_stromal <- subset(mouse_carotid,idents = c(0,2,1,7))
mouse_carotid_stromal <- mouse_carotid_stromal %>% RunPCA() %>% FindNeighbors(dims = 1:20) %>% 
    RunUMAP(dims = 1:20) %>% FindClusters(resolution = 0.1)
mouse_carotid_stromal <- mouse_carotid_stromal %>% FindClusters(resolution = 0.2)

mouse_carotid_stromal <- readRDS("mouse_carotid_stromal.rds")
# saveRDS(mouse_carotid_stromal,"mouse_carotid_stromal.rds")

umapplot(mouse_carotid_stromal)
f("Dlx2",mouse_carotid_stromal) #Dlx2,Dlx5,Dlx6共同定位

mouse_carotid_stromal <- AddModuleScore(mouse_carotid_stromal, list(mmSMC2_marker), name = "SMC2_score")
f("SMC2_score1", mouse_carotid_stromal,min.cutoff = 0)

multi_featureplot(mmSMC2_marker[1:9],mouse_carotid_stromal,labels = NA, label = F)
```

#### 分群表
```{r fig.width=4, fig.height=3}
Dlx5poscells <- WhichCells(mouse_carotid_stromal, expression = `Dlx6` > 0 & `Dlx5` > 0)
group_tab <- Idents(mouse_carotid_stromal)
table(group_tab[Dlx5poscells])/table(group_tab) 
# 在cluster5 有43.8%的细胞表达DLX5，38.4%的细胞表达DLX6,20.2%的细胞表达两者

## 关于样本信息
Dlx5poscells <- WhichCells(mouse_carotid_stromal, idents = "DLX SMC")
group_tab <- mouse_carotid_stromal$orig.ident
table(group_tab[Dlx5poscells])/table(group_tab) 
table(group_tab[Dlx5poscells])
```

#### 图
```{r fig.width=4, fig.height=3}
levels(Idents(mouse_carotid_stromal)) <- c("SEM cell","Fibroblast1","SMC","SMC","Fibroblast2","DLX SMC","Unannotated","Unannotated")
umapplot(mouse_carotid_stromal,label.size = 6,label = T) %>% ggsave("./fig7_mouse/cir_mouse_carotid_stromalumap2.png",plot = ., height = 5, width = 6,device = png)

p <- multi_featureplot(c("Acta2","Cnn1","Fn1","Lum","Vcam1","Ly6a","Dlx5","Dlx6","Lgals3"),mouse_carotid_stromal,labels = NA, label.size = 6, label = F)
ggsave("./fig7_mouse/cir_mouse_carotid_stromal2.png" ,plot = p, height = 12, width = 16,device = png)
Dotplot(c("Acta2","Cnn1","Fn1","Lum","Vcam1","Ly6a","Dlx5","Dlx6","Sost"),mouse_carotid_stromal)

DLXposSMC_markers <- FindMarkers(mouse_carotid_stromal, ident.1 = "DLX SMC", logfc.threshold = 0.4, only.pos = T, min.diff.pct = 0.2)
f("Dlx5",mouse_carotid_stromal)

library(org.Mm.eg.db)
GO_dotplot(rownames(DLXposSMC_markers), OrgDb = org.Mm.eg.db)
```


#### XGBoost 跨物种
```{r}
library(homologene)
mouse_gene2human_gene <- function(genelist){
  mmgenes <- homologene(genelist, inTax = 10090, outTax = 9606)
  mmgenes <- mmgenes[!duplicated(mmgenes$`10090`),]
  genelist[genelist %in% mmgenes$`10090`] <- mmgenes$`9606`
  return(genelist)
}
mouse_carotid_stromal@assays[["SCT"]]@data@Dimnames[[1]] <-
  mouse_carotid_stromal@assays[["SCT"]]@data@Dimnames[[1]] %>% mouse_gene2human_gene()

mouse_carotid_stromal@assays[["SCT"]]@var.features <- mouse_carotid_stromal@assays[["SCT"]]@var.features %>% mouse_gene2human_gene()

f("DLX6", mouse_carotid_stromal)

mouse_carotid_stromal <- query_scmap_from_refsce(mouse_carotid_stromal, ref_sce)

# run xgboost wrapper
mouse_carotid_stromal <- XGBoost_predict_from_seuobj(mouse_carotid_stromal, bst_model, 
                                                     celltype_assign = 2)
mouse_carotid_stromal <- project2ref_celltype(mouse_carotid_stromal, ds2)
umapplot(mouse_carotid_stromal, group.by = "ref_celltype")
umapplot(mouse_carotid_stromal, group.by = "scmap_idents")
table(mouse_carotid_stromal$ref_celltype)
```


```{r}
mousecor_stromal@assays[["SCT"]]@data@Dimnames[[1]] <-
  mousecor_stromal@assays[["SCT"]]@data@Dimnames[[1]] %>% mouse_gene2human_gene()

mousecor_stromal@assays[["SCT"]]@var.features <- mousecor_stromal@assays[["SCT"]]@var.features %>% mouse_gene2human_gene()

f("DLX6", mousecor_stromal)

mousecor_stromal <- query_scmap_from_refsce(mousecor_stromal, ref_sce)

# run xgboost wrapper
mousecor_stromal <- XGBoost_predict_from_seuobj(mousecor_stromal, bst_model, 
                                                     celltype_assign = 2)
mousecor_stromal <- project2ref_celltype(mousecor_stromal, ds2)
umapplot(mousecor_stromal, group.by = "ref_celltype")
umapplot(mousecor_stromal, group.by = "scmap_idents")
table(mousecor_stromal$ref_celltype)
```

# human ds2
```{r}
p <- multi_featureplot(c("ACTA2","CNN1","FN1","LUM","VCAM1","LY6A","DLX5","DLX6","LGALS3","SOST"), ds2, labels = NA, label.size = 6)

ggsave("refds2_carotid_stromal.png",plot = p, height = 12, width = 12,device = png)

p <- multi_featureplot(c("FRZB","SOST","DLX5","DLX6"), ds2, labels = NA, label.size = 6)
ggsave("refds2_SMC2_carotid_stromal.png",plot = p, height = 7, width = 7,device = png)
```


# human bulk RNA-seq *GSE120521* carotid stable/unstable FPKM
```{r}
fpkm2tpm <- function(fpkm){
  exp(log(fpkm) - log(sum(fpkm)) + log(1e6))
}

fpkm_matrix <- read.csv("GSE120521_FPKM.csv")
# fpkm_matrix <- distinct(fpkm_matrix) #去除重复行
fpkm_matrix <- fpkm_matrix[!duplicated(fpkm_matrix$name),]
rownames(fpkm_matrix) <- fpkm_matrix$name
fpkm_matrix$name <- NULL

tpm_matrix <- apply(fpkm_matrix, 2, fpkm2tpm)
colSums(tpm_matrix)

group_file <- c("stable","unstable","stable","unstable",
                "stable","unstable","stable","unstable")
boxplot(tpm_matrix, las = 2)

expr_mat <- tpm_matrix[!apply(tpm_matrix, 1, function(x){sum(floor(x) == 0)>3}),]

boxplot(expr_mat, las = 2)

library(limma)
expr_mat <- normalizeBetweenArrays(expr_mat)
expr_mat <- log2(expr_mat+1) #使用log2 scale

#PCA
library(ggfortify) 
df <- as.data.frame(t(expr_mat)) 
df$group <- group_file 
autoplot(prcomp(df[,1:(ncol(df)-1)]), data=df, colour = 'group')+ theme_bw() 

# If the sequencing depth is reasonably consistent across the RNA samples, then the simplest and most robust approach to differential exis to use limma-trend.

fit <- lmFit(expr_mat, group_file)
fit <- treat(fit, lfc=log2(1.2), trend=TRUE)
topTreat(fit, coef=ncol(design))

```


```{r}
library(ggpubr)
dat <- expr_mat
design <- model.matrix(~factor(group_file))
fit <- lmFit(dat, design)
fit <- eBayes(fit)
# options(digits = 4)
topTable(fit,coef=2,adjust='BH')
deg <- topTable(fit,coef=2,adjust='BH',number = Inf)
head(deg) 

write.csv(deg,"./datatable/stable vs unstable.csv")
## look up FRZB, SOST, PRDM6, OGN
ds1markers[ds1markers$cluster == "SMC2",]$gene

ds2markers[ds2markers$cluster == "SMC3",]$gene

smc2markers <- intersect(ds1markers[ds1markers$cluster == "SMC2",]$gene, ds2markers[ds2markers$cluster == "SMC3",]$gene)


deg[intersect(smc2markers, rownames(deg)),]

ds1markers[ds1markers$cluster == "SMC1",]$gene

ds2markers[ds2markers$cluster == "SMC1",]$gene

SMC1markers <- intersect(ds1markers[ds1markers$cluster == "SMC1",]$gene, ds2markers[ds2markers$cluster == "SMC1",]$gene)

deg[intersect(SMC1markers, rownames(deg)),]
```

```{r}
# logFC_threshold = 1
# pvalue_threshold = 0.05
selected_genes = as.character(read.table("SMC2")$V1)

volcano_plot <- function(filename, selected_genes, logFC_threshold = 1, pvalue_threshold = 0.05)
{
  f<-read.csv("./datatable/stable vs unstable.csv")
  f$threshold <- factor(ifelse(f$adj.P.Val < pvalue_threshold & abs(f$logFC) >= logFC_threshold, 
                              ifelse(f$logFC>= logFC_threshold ,'Up','Down'),'N.S.'),
                       levels=c('Up','Down','N.S.'))
  
   ggplot(f,aes(x=logFC,y=-log10(adj.P.Val),color=threshold))+
    geom_point()+
    scale_color_manual(values=c("#CC0000","#2f5688","#BBBBBB"))+
    geom_text_repel(
      data = f[f$X %in% selected_genes,],
      aes(label = X),
      size = 5, max.overlaps = 1000,
      col="black", segment.color = "black", show.legend = FALSE )+
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          panel.background = element_blank(), axis.line = element_line(colour = "black"))+
    theme(
      axis.title.x = element_text(size = 20), axis.text.x = element_text(size = 15),
      axis.title.y = element_text(size = 20), axis.text.y = element_text(size = 15), 
      legend.text = element_text(size = 20), 
      legend.title = element_blank()
    )+
    ylab('-log10 (p-adj)') +
    xlab('log2 (FoldChange)') +
    geom_vline(xintercept=c(-logFC_threshold,logFC_threshold), lty=3,col="black",lwd=0.5) +
    geom_hline(yintercept =c(0,-log10(pvalue_threshold)),lty=3,col="black",lwd=0.5)
}

selected_genes <- c(as.character(read.table("SMC2")$V1))

p <- volcano_plot("./datatable/stable vs unstable.csv",selected_genes) + ggtitle("stable vs unstable, SMC2 marker")

ggsave("SMC2_stable vs unstable.png",plot = p,device = png,height = 4, width = 6)
```


Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.
